src/aesgcm.rs

```diff
@@ -0,0 +1,261 @@
+use error::{ErrorKind, Result};
+
+const ECE_AESGCM_PAD_SIZE: usize = 2;
+
+const ECE_WEBPUSH_AESGCM_KEYPAIR_LENGTH: usize = 134; // (2 + 65) * 2
+const ECE_WEBPUSH_AESGCM_AUTHINFO: &'static str = "Content-Encoding: auth\0";
+
+// const ECE_WEBPUSH_DEFAULT_RS: u32 = 4096;
+// a DER prefixed key is "\04" + ECE_WEBPUSH_RAW_KEY_LENGTH
+const ECE_WEBPUSH_RAW_KEY_LENGTH: usize = 65;
+const ECE_WEBPUSH_IKM_LENGTH: usize = 32;
+
+pub struct AesGcmEncryptedBlock {
+    pub dh: Vec<u8>,
+    pub salt: Vec<u8>,
+    pub rs: u32,
+    pub ciphertext: Vec<u8>,
+}
+
+impl AesGcmEncryptedBlock {
+    pub fn aesgcm_rs(rs: u32) -> u32 {
+        if rs > u32::max_value() - ECE_TAG_LENGTH as u32 {
+            return 0;
+        }
+        rs + ECE_TAG_LENGTH as u32
+    }
+
+    /// Create a new block from the various header strings and body content.
+    pub fn new(dh: Vec<u8>, salt: Vec<u8>, rs: u32, ciphertext: Vec<u8>) -> Result<AesGcmEncryptedBlock> {
+        Ok(AesGcmEncryptedBlock {
+            //dh: base64::decode_config(&dh, base64::URL_SAFE_NO_PAD)?,
+            //salt: base64::decode_config(&salt, base64::URL_SAFE_NO_PAD)?,
+            dh,
+            salt,
+            rs: Self::aesgcm_rs(rs),
+            ciphertext,
+        })
+    }
+
+    /// Return the headers Hash, NOTE you may need to merge Crypto-Key if there's
+    /// already a VAPID element present.
+    pub fn headers(self) -> HashMap<String, String> {
+        let mut result: HashMap<String, String> = HashMap::new();
+        let mut rs = "".to_owned();
+        result.insert(
+            "Crypto-Key".to_owned(),
+            format!(
+                "dh={}",
+                base64::encode_config(&self.dh, base64::URL_SAFE_NO_PAD)
+            ),
+        );
+        if self.rs > 0 {
+            rs = format!(";rs={}", self.rs);
+        }
+        result.insert(
+            "Encryption".to_owned(),
+            format!(
+                "salt={}{}",
+                base64::encode_config(&self.salt, base64::URL_SAFE_NO_PAD),
+                rs
+            ),
+        );
+        result
+    }
+
+    /// Encode the body as a String.
+    /// If you need the bytes, probably just call .ciphertext directly
+    pub fn body(self) -> String {
+        base64::encode_config(&self.ciphertext, base64::URL_SAFE_NO_PAD)
+    }
+}
+
+pub struct AesGcmEceWebPush<L, R, C> {
+    _marker1: ::std::marker::PhantomData<L>,
+    _marker2: ::std::marker::PhantomData<R>,
+    _marker3: ::std::marker::PhantomData<C>,
+}
+
+impl<L, R, C> AesGcmEceWebPush<L, R, C>
+where
+    L: LocalKeyPair,
+    R: RemotePublicKey,
+    C: Crypto<LocalKeyPair = L, RemotePublicKey = R>,
+{
+    /// Encrypts a Web Push message using the "aesgcm" scheme. This function
+    /// automatically generates an ephemeral ECDH key pair.
+    pub fn encrypt(
+        remote_pub_key: &R,
+        auth_secret: &[u8],
+        plaintext: &[u8],
+        params: WebPushParams,
+    ) -> Result<AesGcmEncryptedBlock> {
```

Comment:
I'm not sure. `aesgcm` is a multi-part thing, more so than `aes128gcm`. Tossing things into the block makes the encrypt/decrypt more symmetric, but does come at the price that both aren't quite the same. I'm not sure how we'd return, say, the ephemeral `auth` and `salt`, which need to go into headers. And since the encrypted block is composed differently as well, I don't think we could do the same composition trick that `aes128gcm` does and then extract the values from the header block.